### PR TITLE
Allow replacing path without replacing query or fragment

### LINKED ISSUE #111: No way to replace path without replacing query
The following is unexpected behavior:

```
url1 = URL("http://www.example.com/foo?bar=10")
url2 = url1.with_path("/bar")

assert url1.query = url2.query  # Assertion fails
```

While I understand the design decision to treat the query and fragment components as part of the path, this has left us without an (efficient) way to preserve the query and fragment components when replacing just the path. The only way to do it is:

```
url2 = url1.with_path(new_path).with_query(url1.query_string).with_fragment(url1.fragment)
```

This is a total of 4 extra method calls, at least of which does a lot of unnecessary processing. I'm assuming that there's code out there that depends on this weird behavior, so I'd like to propose and, with permission, implement a new method, `with_just_path`, which replaces the path but leaves the fragment and query the same.

### PR DESCRIPTION
**Pull request by paul-nameless**
## What do these changes do?

Add keep_query and keep_fragment arguments to with_path function to allow replacing path without  replacing query or fragment

## Are there changes in behavior for the user?

Yes, users can now replace the path in a URL without unintentionally modifying the query string or fragment by using the new arguments. Existing behavior remains unchanged unless the new arguments are explicitly used.

## Related issue number

closes #111 

## Checklist

- [X] I think the code is well written
- [X] Unit tests for the changes exist
- [ ] Documentation reflects the changes

### CODE REVIEW

File: CHANGES/111.bugfix.rst
Comment (asvetlov) [inline CHANGES/111.bugfix.rst:1]: Please update the text. Something like "add `keep_query` and `keep_fragment` flags to <enumerate affected methods> for optional keeping query string and fragment in the resulting URL". The exact text is up to you but please keep in mind: it would be nice if a user could get a brief idea of the change without digging into PR content.

Comment (paul-nameless) [reply CHANGES/111.bugfix.rst]: Sure, it makes more sense.
Done

File: yarl/_url.py
Comment (bdraco) [inline yarl/_url.py:1128]:
```suggestion
        keep_query: bool = False,
        keep_fragment: bool = False,
```

File: yarl/_url.py
Comment (bdraco) [inline yarl/_url.py:1139]:
```suggestion
        query = self._query if keep_query else ""
        fragment = self._fragment if keep_fragment else ""
```

File: tests/test_url.py
Comment (bdraco) [inline tests/test_url.py:1240]:
```suggestion
    url2 = url.with_path("/test", keep_query=True)
```

File: tests/test_url.py
Comment (bdraco) [inline tests/test_url.py:1257]:
```suggestion
    url2 = url.with_path("/test", keep_query=True, keep_fragment=True)
```

File: tests/test_url.py
Comment (bdraco) [inline tests/test_url.py:1259]: Please add tests for explicitly passing `keep_fragment=False` and `keep_query=False`

File: CHANGES/111.bugfix.rst
Comment (bdraco) [inline CHANGES/111.bugfix.rst:1]:
```suggestion
Added ``keep_query`` and ``keep_fragment`` flags in the :py:meth:`yarl.URL.with_path`, allowing users to optionally retain the query string and fragment in the resulting URL when replacing the path -- by :user:`paul-nameless`.
```

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
reframed this as a new feature since the current behavior doesn't change and its by design. All the functionality here is new.

**Comment by bdraco** [comment]
Screenshot 2024-11-18 at 12 36 02 PM

**Comment by bdraco** [comment]
One last thing, we also need to update the docs for `with_path` with the new params

https://yarl--1421.org.readthedocs.build/en/1421/api/#yarl.URL.with_path

**Comment by asvetlov** [comment]
We also have `.with_name()` and `.with_suffix()`.
Maybe these methods need proposed skip options as well?

**Comment by paul-nameless** [comment]
Updated the documentation

**Review by bdraco — APPROVED** [review]

File: CHANGES/111.feature.rst
Comment (asvetlov) [inline CHANGES/111.feature.rst:1]: Sorry, I've missed. How is this file rendered finally? What is desired CHANGELOG record?

File: docs/api.rst
Comment (asvetlov) [inline docs/api.rst:662]:
```suggestion
   If ``keep_query=True`` or ``keep_fragment=True`` it retains the existing query or fragment in the URL.

   .. versionchanged:: 1.18

      Added *keep_query* and *keep_fragment* parameters.
```

File: CHANGES/111.feature.rst
Comment (bdraco) [reply CHANGES/111.feature.rst:1]: Screenshot 2024-11-18 at 9 28 03 PM

https://yarl--1421.org.readthedocs.build/en/1421/changes/#features

Comment (bdraco) [reply CHANGES/111.feature.rst:1]: Usually I use the PR# for the changelog messages. If the issue has a lot in it, symlink to the PR as well so both show up

File: docs/api.rst
Comment (bdraco) [inline docs/api.rst:664]:
```suggestion

   .. doctest::
```

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
We will also need to bump the version to 1.18.0.dev0 after merging this PR

File: CHANGES/111.feature.rst
Comment (asvetlov) [reply CHANGES/111.feature.rst:1]: Got you

**Comment by asvetlov** [comment]
> We will also need to bump the version to 1.18.0.dev0 after merging this PR

Technically, we can do it right now.

File: tests/test_url.py
Comment (webknjaz) [inline tests/test_url.py:1264]: Looks like these could be collapsed via `@pytest.mark.parametrize()`. Plus, there's the case of `keep_query=False` with `keep_fragment=False` that seems to be missing, but could be added as another param.

File: tests/test_url.py
Comment (webknjaz) [inline tests/test_url.py:1250]: Let's add context by naming each param:
```suggestion
    ("original_url", "keep_query", "keep_fragment", "expected_url"),
    (
        pytest.param("http://example.com?a=b#frag", True, False, "http://example.com/test?a=b", id="query-only"),
        pytest.param("http://example.com?a=b#frag", False, True, "http://example.com/test#frag", id="fragment-only"),
        pytest.param("http://example.com?a=b#frag", True, True, "http://example.com/test?a=b#frag", id="all"),
        pytest.param("http://example.com?a=b#frag", False, False, "http://example.com/test", id="none"),
    ),
```

File: tests/test_url.py
Comment (webknjaz) [inline tests/test_url.py:16]: Seeing that the same params are being used in several tests, let's just save the marker into a variable:
```suggestion

with_query_frag_preservation_matrix = pytest.mark.parametrize(
    ("original_url", "keep_query", "keep_fragment", "expected_url"),
    (
        pytest.param("http://example.com?a=b#frag", True, False, "http://example.com/test?a=b", id="query-only"),
        pytest.param("http://example.com?a=b#frag", False, True, "http://example.com/test#frag", id="fragment-only"),
        pytest.param("http://example.com?a=b#frag", True, True, "http://example.com/test?a=b#frag", id="all"),
        pytest.param("http://example.com?a=b#frag", False, False, "http://example.com/test", id="none"),
    ),
)


```

And then, use it on the tests down below as follows:
```python
@with_query_frag_preservation_matrix
def test_...(
    original_url: str,
    keep_query: bool,
    keep_fragment: bool,
    expected_url: str,
): -> None:
    ...
```

Comment (asvetlov) [reply tests/test_url.py:16]: I'm -0 for this proposal.
It reduces tests *redundancy* but on the other hand it reduces the tests *readability* as well; I should look at the variable content to figure out that parameters are used.
As a maintainer, I read the code more often than write it.
I like a moderate portion of redundancy in tests, it simplifies reading and debugging at least in *my personal* workflow. Other people could have an opposite opinion, sure.

File: tests/test_url.py
Comment (asvetlov) [reply tests/test_url.py:1250]: Agree, the proposed change makes tests much better readable.

File: yarl/_url.py
Comment (asvetlov) [inline yarl/_url.py:1293]:
```suggestion
    def with_name(
        self,
        name: str,
        *,
        keep_query: bool = False,
        keep_fragment: bool = False,
    ) -> "URL":
```

File: yarl/_url.py
Comment (asvetlov) [inline yarl/_url.py:1330]:
```suggestion
    def with_suffix(
        self,
        suffix: str,
        *,
        keep_query: bool = False,
        keep_fragment: bool = False,
    ) -> "URL":
```

**Review by asvetlov — CHANGES_REQUESTED** [review]
Please update affected method signatures in docs to reflect added parameters.

File: docs/api.rst
Comment (asvetlov) [inline docs/api.rst:655]:
```suggestion
.. method:: URL.with_path(path, *, keep_query=False, keep_fragment=False)
```

File: docs/api.rst
Comment (asvetlov) [inline docs/api.rst:866]:
```suggestion
.. method:: URL.with_name(name, *, keep_query=False, keep_fragment=False)
```

File: docs/api.rst
Comment (asvetlov) [inline docs/api.rst:886]:
```suggestion
.. method:: URL.with_suffix(suffix, *, keep_query=False, keep_fragment=False)
```

**Review by asvetlov — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
